File: docs/manual/guide/cluster/code/docs/home/persistence/RDBMSReadSideQuery.java
package docs.home.persistence;

//#imports

import akka.NotUsed;
import com.lightbend.lagom.javadsl.api.ServiceCall;
import com.lightbend.lagom.javadsl.persistence.jdbc.JdbcSession;
import org.pcollections.PSequence;
import org.pcollections.TreePVector;

import javax.inject.Inject;
import java.sql.ResultSet;
//#imports

public interface RDBMSReadSideQuery {

    interface BlogService {
        public ServiceCall<NotUsed, PSequence<PostSummary>> getPostSummaries();
    }

    //#service-impl
    public class BlogServiceImpl implements BlogService {

        private final JdbcSession jdbcSession;

        @Inject
        public BlogServiceImpl(JdbcSession jdbcSession) {
            this.jdbcSession = jdbcSession;
        }

        @Override
        public ServiceCall<NotUsed, PSequence<PostSummary>> getPostSummaries() {
            return request -> {
                return jdbcSession.withConnection(connection -> {
                    ResultSet rs = connection.prepareStatement("SELECT id, title FROM blogsummary")
                            .executeQuery();
                    PSequence<PostSummary> summaries = TreePVector.empty();

                    while (rs.next()) {
                        summaries = summaries.plus(
                                new PostSummary(rs.getString("id"), rs.getString("title"))
                        );
                    }

                    return summaries;
                });
            };
        }
    }
    //#service-impl

}

File: docs/manual/scala/guide/cluster/code/docs/home/scaladsl/persistence/RDBMSReadSideQuery.scala
package docs.home.scaladsl.persistence

//#imports
import scala.collection.immutable
import scala.collection.immutable.VectorBuilder
import akka.NotUsed
import com.lightbend.lagom.scaladsl.api.Service
import com.lightbend.lagom.scaladsl.api.ServiceCall
import com.lightbend.lagom.scaladsl.persistence.jdbc.JdbcSession

//#imports

trait RDBMSReadSideQuery {

  trait BlogService extends Service {
    def getPostSummaries(): ServiceCall[NotUsed, immutable.IndexedSeq[PostSummary]]

    override def descriptor = ???
  }

  //#service-impl
  class BlogServiceImpl(jdbcSession: JdbcSession) extends BlogService {

    override def getPostSummaries() = ServiceCall { request =>
      jdbcSession.withConnection { connection =>
        val rs = connection.prepareStatement("SELECT id, title FROM blogsummary")
          .executeQuery()
        val summaries = new VectorBuilder[PostSummary]
        while (rs.next()) {
          summaries +=
            PostSummary(rs.getString("id"), rs.getString("title"))
        }
        summaries.result()
      }
    }
    //#service-impl

  }
}
